### PR TITLE
Add GradientReversal layer

### PR DESCRIPTION
**Pull request by Kaixhin**
Implemented a GradientReversal layer, as seen in [Domain-Adversarial Training of Neural Networks](http://arxiv.org/abs/1505.07818) (though quite possibly not its first appearance). Can be used to maximise a loss function with gradient descent, especially in tandem with minimising other loss functions normally.

Also did a little tidying up of the docs.

### CODE REVIEW

**Comment by soumith** [comment]
kai, can you add simple tests? even trivial ones will do.

**Comment by soumith** [comment]
you also have to add it to init.lua

**Comment by Kaixhin** [comment]
Done! I would echo Andrej's [comment](https://gitter.im/torch/torch7/archives/2015/05/17) on adding your tips for development to [CONTRIBUTING.md](https://github.com/torch/nn/blob/master/CONTRIBUTING.md), seeing as I dug it up trying to get the tests working.

**Comment by soumith** [comment]
Thanks a lot Kai!!!

**Comment by soumith** [comment]
Added the tips to the CONTRIBUTING doc: https://github.com/torch/nn/blob/master/CONTRIBUTING.md#development-workflow-tips

**Comment by Kaixhin** [comment]
Cheers - I'm sure it'll help some potential contributors in the future.
